x-pack/legacy/plugins/infra/server/lib/alerting/metric_threshold/register_metric_threshold_alert_type.ts

```diff
@@ -0,0 +1,143 @@
+              },
+            },
+          },
+          {
+            bool: {
+              should: [
+                {
+                  match_phrase: {
+                    [searchField.name]: searchField.value,
+                  },
+                },
+                {
+                  exists: {
+                    field: metric,
+                  },
+                },
+              ],
+            },
+          },
+        ],
+      },
+    },
+    size: 0,
+    aggs: {
+      aggregatedIntervals: {
+        date_histogram: {
+          field: '@timestamp',
+          calendar_interval: interval,
+        },
+        aggregations: {
+          aggregatedValue: {
+            [aggregation]: {
+              field: metric,
+            },
+          },
+        },
+      },
+    },
+  };
+
+  const result = await callCluster('search', {
+    body: searchBody,
+    index: indexPattern,
+  });
+
+  const { buckets } = result.aggregations.aggregatedIntervals;
+  const { value } = buckets[buckets.length - 1].aggregatedValue;
+  return value;
+}
+
+const comparatorMap = {
+  [Comparator.GT]: (a: number, b: number) => a > b,
+  [Comparator.LT]: (a: number, b: number) => a < b,
+  [Comparator.GT_OR_EQ]: (a: number, b: number) => a >= b,
+  [Comparator.LT_OR_EQ]: (a: number, b: number) => a <= b,
+};
+
+export async function registerMetricThresholdAlertType(server: Server) {
+  const { alerting } = server.plugins;
+  if (!alerting) {
+    throw new Error(
+      'Cannot register metric threshold alert type.  Both the actions and alerting plugins need to be enabled.'
+    );
+  }
+
+  alerting.setup.registerType({
+    id: METRIC_THRESHOLD_ALERT_TYPE_ID,
+    name: 'Metric Alert - Threshold',
+    validate: {
+      params: schema.object({
+        threshold: schema.number(),
+        comparator: schema.string(),
+        aggregation: schema.string(),
+        searchField: schema.object({
+          name: schema.string(),
+          value: schema.string(),
+        }),
+        metric: schema.string(),
+        interval: schema.string(),
+        indexPattern: schema.string(),
+      }),
+    },
+    actionGroups: [FIRED_ACTIONS],
+    async executor({ services, params }) {
+      const { threshold, comparator, searchField } = params as MetricThresholdAlertTypeParams;
+      const alertInstance = services.alertInstanceFactory(
+        `${searchField.name}:${searchField.value}`
+      );
+      const currentValue = await getMetric(services, params as MetricThresholdAlertTypeParams);
+      if (typeof currentValue === 'undefined')
+        throw new Error('Could not get current value of metric');
+
+      const comparisonFunction = comparatorMap[comparator];
+
+      const isValueInAlertState = comparisonFunction(currentValue, threshold);
+
+      if (isValueInAlertState) {
+        alertInstance.scheduleActions(FIRED_ACTIONS, {
+          value: currentValue,
+        });
+      }
+
+      alertInstance.replaceState({
```

Comment:
You don't seem to be using this alertInstance state being stored at this point, anywhere else (at least in the alert), so you technically don't need this.  Perhaps it's for a future use ...

Reply:
I figured that we'd need this for a UI that displays alerts and their current states/historical states. I know that's not implemented yet but I assume it'll use the `api/alert/<id>/state` API? If it's still up in the air I can remove this and we'll handle that later.

Reply:
The alertInstance state is typed as a `Record<string, unknown>`, and as such, the alerting plugin itself is unlikely to expose any generic UI that could display it.  A consumer-specific plugin certainly could.  I think it's fine to leave this in, perhaps add a comment that this could be retrieved by a metrics plugin to provide alert-instance specific data in a UI.  ie, "future use"